### ISSUE TITLE
all: umbrella issue for release-blocking fuzzing work

### ISSUE THREAD

**Issue by katiehockman** [golang/go#47037]
This is the tracking issue for fuzzing-related work which must be done before fuzzing can be part of a standard release of the Go toolchain.

The current plan is to merge the fuzzing code into master once the tree opens in the fall of this year (2021).
**This would make fuzzing available for Go 1.18**.

The following is a list of release-blocking features (in no particular order):

- [x] Merge fuzzing into master, removing the default build tag (gofuzzbeta) currently in the dev.fuzz branch
- [x] (can be during freeze) Documentation describing which types are currently supported for fuzzing
- [x] A decision must be made on whether or not fuzzing will support generics, as adding this later would break backwards compatibility for existing fuzz targets
  - #46890
- [x] Minimization of interesting inputs
- [x] A decision must be made for how `-cpu` is used
  - https://github.com/golang/go/issues/46629
- [x] A decision must be made on whether other tests run when `-fuzz` is set
  - https://github.com/golang/go/issues/46222
- [x] A decision must be made on whether types can be converted when using `f.Add` and `f.Fuzz` (may decide to make this more strict for now)
  - https://github.com/golang/go/issues/45593
- [x] Don't store corpus in memory
  - https://github.com/golang/go/issues/46669
- [ ] Fuzzing works with -race and -msan. It doesn't have to add a lot of additional value, but the go tool shouldn't break.

The following is a list of nice-to-have features (in a rough order):
- [ ] Detect timeouts and provide a `-timeout` flag.
  - https://github.com/golang/go/issues/48157
- [ ] Allow multiple fuzz targets to be run with a single go command
  - #46312
- [ ] The mutators might be a bit too aggressive, and could be improved (https://github.com/golang/go/issues/47090)
- [x] A tool that can convert binary files into the fuzz encoding file format
  - Related to https://github.com/golang/go/issues/46311#issue-898218779
- [ ] Implement better deduplication of interesting inputs so the ondisk corpus doesn't get too inflated
- [ ] Deduplication of crashes
- [ ] Support for -keepfuzzing (This is contingent on deduplication working first)
- [ ] Support for fuzzing more builtin types, such as slices and arrays
- [ ] Supporting dictionaries
- [ ] Support for fuzzing structs
- [ ] Prioritization of newly discovered interesting inputs, and perhaps some minimal work done on newly discovered interesting inputs.
- [ ] Improved mutator - specifically one that uses other corpus entries while mutating
- [ ] Supports a build mode which can generate a compiled binary to be run using libFuzzer

There are other issues with the 'fuzz' label which will also be release blockers once fuzzing is merged. Once the tree is open, the release blocker label will be used at that time.

Which items are in the "release-blocker" list and which are in the "nice-to-have" list may change over time, depending on user feedback, and new items may be added at any time as well.

/cc @golang/fuzzing

**Comment by AlekSi** [comment]
> Fuzzing works with -race and -msan. It doesn't have to add a lot of additional value, but the go tool shouldn't break.

Currently, `-race` reports a data race in `SnapshotCoverage()`. Is it a known problem? Should there be a separate ticket for that?

**Comment by AlekSi** [comment]
(Also, this ticket is a bit confusing. For example, the title says "release-blocking", but the milestone is "Unreleased")

**Comment by katiehockman** [comment]
I went ahead and updated the milestone. (This is just an umbrella issue used for tracking and external visibility, so any issues/requests should be filed as separate tickets)

@AlekSi could you file a new issue for the `-race` issue. We need to look into that again, and AFAIK there isn't an open ticket for it. Thanks.

**Comment by katiehockman** [comment]
@rolandshoemaker tried to reproduce that race issue locally, and wasn't able to. I'm going to go ahead and mark that done in the umbrella issue, and just close this. This issue was really just for tracking anyway.

But @AlekSi please still file your issue if you can reproduce it at tip.

**Comment by thepudds** [comment]
FWIW, I’ve been using -race and -fuzz together a fair amount for the past several weeks without an issue, so maybe it is related to the exact code under test.

**Comment by AlekSi** [comment]
#50488

**Comment by katiehockman** [comment]
Thanks, @AlekSi for filing that issue!
